File: spcon/public/py/sales_invoice.py
import frappe
from frappe.utils import formatdate


def get_sales_order_remark_text(sales_orders):
    sales_orders = list(dict.fromkeys(filter(None, sales_orders or [])))

    if not sales_orders:
        return ""

    sales_order_dates = frappe.get_all(
        "Sales Order",
        filters={"name": ["in", sales_orders]},
        fields=["name", "transaction_date", "po_date"],
    )
    date_map = {row.name: row.po_date for row in sales_order_dates}

    remarks = []
    for sales_order in sales_orders:
        sales_order_date = date_map.get(sales_order)
        formatted_date = formatdate(sales_order_date, "dd-mm-yyyy") if sales_order_date else ""
        # remarks.append(f"({sales_order} = {formatted_date})")
        remarks.append(f"({formatted_date})")

    return ", ".join(remarks)


@frappe.whitelist()
def get_sales_order_remark(sales_orders):
    if isinstance(sales_orders, str):
        sales_orders = frappe.parse_json(sales_orders)

    return get_sales_order_remark_text(sales_orders)


@frappe.whitelist()
def set_sales_order_remark(doc, method=None):
    doc.custom_remark = get_sales_order_remark_text(
        item.sales_order for item in doc.items
    )


@frappe.whitelist() 
def set_actual_dispatch_date_on_save(doc, method=None):
    for item in doc.items:
        
        if item.sales_order and item.so_detail:
            
            frappe.db.set_value(
                "Sales Order Item",        
                item.so_detail,            
                "custom_actual_dispatch_date",
                doc.posting_date           
            )

@frappe.whitelist()
def set_minimum_qty(doc, method=None):
    errors = []

    if doc.items and doc.is_return != 1:
        for item in doc.items:
            min_qty = frappe.get_value("Item", item.item_code, "custom_minimum_sale_qty") or 0

            if item.qty < min_qty:
                errors.append(
                    f"Row {item.idx}: Item {item.item_code} → Minimum Qty = {min_qty}, Entered Qty = {item.qty}"
                )

    if errors:
        # frappe.throw("<br>".join(errors))
        frappe.msgprint("<br>".join(errors))


@frappe.whitelist()
def validate_naming_series(doc, method=None):

    data = frappe.get_all(
        "Cost Center Naming Series",
        {"parent": doc.cost_center},
        ["naming_series"]
    )

    allowed_series = [d.naming_series for d in data]

    if doc.naming_series not in allowed_series:
        frappe.throw(
            f"Invalid naming series for the selected cost center."
        )

File: spcon/public/js/sales_invoice.js
frappe.ui.form.on('Sales Invoice', {
	async cost_center(frm) {
		if (!frm.doc.cost_center) {
			frm.set_value('company_address', '');
			return;
		}

		const r = await frappe.db.get_value('Cost Center', frm.doc.cost_center, 'custom_address');
		const address = r.message?.custom_address || '';
		frm.set_value('company_address', address);
		console.log(address);
		apply_warehouse_filter(frm)
	},
	onload(frm) {
		apply_warehouse_filter(frm);
	},
	refresh(frm) {
		setTimeout(() => set_sales_order_remark(frm), 300);
	},
	get_items(frm) {
		setTimeout(() => set_sales_order_remark(frm), 500);
	},
	async validate(frm) {
		await set_sales_order_remark(frm);
	}
});

frappe.ui.form.on('Sales Invoice Item', {
	items_add(frm) {
		set_sales_order_remark(frm);
	},
	items_remove(frm) {
		set_sales_order_remark(frm);
	},
	sales_order(frm) {
		set_sales_order_remark(frm);
	},
	item_code(frm) {
		setTimeout(() => set_sales_order_remark(frm), 300);
	}
});


function apply_warehouse_filter(frm) {
    if (!frm.doc.cost_center) return;

    frm.set_query("set_warehouse", function() {
        return {
            query: "spcon.public.py.warehouse.get_warehouses_for_cost_center",
            filters: { 
                cost_center: frm.doc.cost_center
            }
        };
    });
}

function set_sales_order_remark(frm) {
	const sales_orders = [
		...new Set((frm.doc.items || []).map((row) => row.sales_order).filter(Boolean))
	];

	if (!sales_orders.length) {
		return frm.set_value('custom_remark', '');
	}

	return frappe.call({
		method: 'spcon.public.py.sales_invoice.get_sales_order_remark',
		args: { sales_orders },
		callback(r) {
			frm.set_value('custom_remark', r.message || '');
		}
	});
}
